resources/waterheater.rb

```diff
@@ -648,51 +649,110 @@ def self.apply_indirect(model, runner, space, cap, vol, t_set, oncycle_p, offcyc
 
     loop.addSupplyBranchForComponent(new_heater)
 
-    # EMS for offsetting reaction lag and recover tank temperature
-    # Sensors
-    use_heat_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Water Heater Use Side Heat Transfer Energy")
-    use_heat_sensor.setName("#{obj_name_indirect} Use Side Energy")
-    use_heat_sensor.setKeyName("#{obj_name_indirect}")
-
-    tank_temp_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Water Heater Tank Temperature")
-    tank_temp_sensor.setName("#{obj_name_indirect} Tank Temp")
-    tank_temp_sensor.setKeyName("#{obj_name_indirect}")
+    dhw_map[sys_id] << add_ec_adj(model, runner, new_heater, ec_adj, space, boiler_fuel_type, "boiler", boiler, indirect_hx)
 
-    stp_temp_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Schedule Value")
-    stp_temp_sensor.setName("#{obj_name_indirect} Setpoint Temperature")
-    stp_temp_sensor.setKeyName("WH Setpoint Temp")
+    return true
+  end
 
-    wh_loss_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Water Heater Heat Loss Energy")
-    wh_loss_sensor.setName("#{obj_name_indirect} Loss Energy")
-    wh_loss_sensor.setKeyName("#{obj_name_indirect}")
+  def self.apply_combi_system_EMS(model, runner, combi_sys_id, dhw_map)
+    # EMS for modulate source side mass flow rate
+    # Initialization
+    equipment_peaks = {}
+    equipment_sch_sensors = {}
+    equipment_target_temp_sensors = {}
+    tank_volume, deadband = 0.0, 0.0
+    alt_spt_sch = nil
+    tank_temp_sensor, tank_spt_sensor, tank_loss_energy_sensor = nil, nil, nil
+    altsch_actuator, pump_actuator = nil, nil
+
+    # Create sensors and actuators by dhw map information
+    dhw_map[combi_sys_id].each do |object|
+      if object.is_a? OpenStudio::Model::WaterUseConnections
+        object.waterUseEquipment.each do |wu|
+          # water use equipment peak mass flow rate
+          wu_peak = wu.waterUseEquipmentDefinition.peakFlowRate
+          equipment_peaks[wu.name.to_s] = wu_peak
+          # mfr fraction schedule sensors
+          wu_sch_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Schedule Value")
+          wu_sch_sensor.setName("#{wu.name.to_s} sch value")
+          wu_sch_sensor.setKeyName(wu.flowRateFractionSchedule.get.name.to_s)
+          equipment_sch_sensors[wu.name.to_s] = wu_sch_sensor
+          # water use equipment target temperature schedule sensors
+          target_temp_sch = wu.waterUseEquipmentDefinition.targetTemperatureSchedule.get
+          target_temp_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, "Schedule Value")
+          target_temp_sensor.setName("#{wu.name.to_s} target temp")
+          target_temp_sensor.setKeyName(target_temp_sch.name.to_s)
+          equipment_target_temp_sensors[wu.name.to_s] = target_temp_sensor
```

Comment:
This code is much cleaner now compared to calling out individual object. Thanks!

Reply:
No problem. :)